### DISCUSSION TITLE
How to call the custom function defined inside another rego

### DISCUSSION THREAD

**Discussion by jiangduzi** [open-policy-agent/community#472] [OPA and Rego] [answered]
Need help, I want to reference different rego files in main.rego based on input.name. Then call custom functions defined inside that rego file but getting errors during compilation "rego_compile_error: called function policy.test shadowed".
My rego just like this:

**main.rego**

```rego
package play
import data.play.policies

allow {
    policy := policies[input.name]
    policy.test(input.params)
}
```

**policy1.rego**

```rego
package play.policies.policy1

test(params) {
    params["env"] == "dev"
}
```

**policy2.rego**

```rego
package play.policies.policy2

test(params) {
    params["env"] == "prod"
}
```

**Comment by srenatus** [comment]
In `main.rego`, try this instead:
```rego
package play
import data.play.policies

allow {
    policy := policies[input.name].test(input.params)
}
```

Before, this is what happened: When you do
```rego
policy := policies[input.name]
```
you'll have OPA evaluate the _full extent_ of `data.play.policies[input.name]` -- that means it'll include all the rules that evaluate to _something_, and all the static data defined in your store under that reference **and none of the functions**. That's because functions don't contribute to the extent, since they wouldn't be called with arguments.

However, the _dynamic dispatch_ thing you want it entirely possible, and that's what should happen in the altered snippet above.

**Reply by jiangduzi** [reply]
OK, I will try it again, thanks

**Reply by jiangduzi** [reply]
@srenatus Sorry, I try it again, but it doesn't work. The build error message is "rego_type_error: undefined function data.play.policies[input.name].test"

The changed main.rego:
```rego
package play
import data.play.policies

allow {
    policies[input.name].test(input.params)
}
```

But if I use this main.rego, it works well.  So I think whether opa can't reference the different rego files based on the variable "input.name"?
```rego
package play
import data.play.policies

allow {
    policies["policy1"].test(input.params)
}
```

**Reply by charlieegan3** [reply]
I've generally seen this done with rules rather than functions, have you tried that? https://www.styra.com/blog/dynamic-policy-composition-for-opa/

**Reply by jiangduzi** [reply]
@charlieegan3 Thanks for the response.
For simplicity, I provided a simple example where I mentioned using functions. In reality, I used functions in order to pass different params in main.rego, so I think rules can't match my requirement.

Just like this:
```rego
package play
import data.play.policies

allow {
    condition := input.conditions[_]
    policies[condition.name].test(condition.params)
}
```

the input:
```json
{
    "input": {
        "conditions": [
            {
                "name": "policy1",
                "params": {
                    "env": "dev"
                }
            },
            {
                "name": "policy2",
                "params": {
                    "env": "prod"
                }
            }  
        ]
    }
}
```

**Reply by charlieegan3** [reply, marked as answer]
Hmm, how about using `with`? https://play.openpolicyagent.org/p/wMu6dvXFUv

**Reply by jiangduzi** [reply]
It seems like a new approach, I have never used "with" before. I will try it tomorrow, Many thanks!
